scipy/linalg/src/_linalg_inv.hh

```diff
@@ -159,6 +159,45 @@ void invert_slice_triangular(
 }
 
 
+// Diagonal array inversion
+template<typename T>
+inline void invert_slice_diagonal(
+    CBLAS_INT N, T *data, SliceStatus& status
+) {
+    using real_type = typename type_traits<T>::real_type;
+    using value_type = typename type_traits<T>::value_type;
+    value_type *pdata = reinterpret_cast<value_type *>(data);
+
+    value_type zero(0.), one(1.);
+    real_type maxa(0.), maxinva(0.);
+
+    for (CBLAS_INT j=0; j<N; j++) {
+        value_type ajj = pdata[j*N + j];
+
+        status.is_singular = (ajj == zero);
+        if (status.is_singular) {
+            status.lapack_info  = j;
+            return;
+        }
+
+        value_type inv_ajj = one / ajj;
+        pdata[j*N + j] = inv_ajj;
+
+        // condition number
+        real_type absa = std::abs(ajj), absinva = std::abs(inv_ajj);
+
+        if(absa != absa) {
```

Comment:
Any reason to prefer this over `std::isnan(absa)`?

And in my usual "is this really tested" stance, if I change this to a codepath that can't be hit `spin test -j 8` still passes in my hands on x86-64 Linux:

```diff
-        if(absa != absa) {
+        if(0) {
             status.is_ill_conditioned = true;
```

Reply:
TIL about `std::isnan`, thank you! 

Writing a test for this branch made me realize it's quite awkward: we check for nans with `check_finite`, so I needed to explicitly switch off the check, and then manually add a nan: `inv([[np.nan, 0], [0, 1], check_finite=False)`. This looks weird, and in the LAPACK-aware solvers we explicitly say that the result is undefined: the exact phrase is _may result in problems (crashes, non-termination) if the inputs do contain infinities or NaNs_. Therefore, the last commit simplifies this branch away. We'll be able to reinstate it if users complain.